Firewalk/man/firewalk.8
.\"   $Id: firewalk.8,v 1.2 2002/05/14 23:28:37 route Exp $: firewalk.1,v 1.1.1.1 2001/03/30 19:14:33 route Exp $
.\"
.\"   Firewalk 5.0
.\"   Firewalk manpage
.\"
.\"   Copyright (c) 1998 - 2002 Mike D. Schiffman  <mike@infonexus.com>
.\"   Copyright (c) 1998, 1999 David E. Goldsmith <dave@infonexus.com>
.\"   http://www.packetfactory.net/firewalk
.\"
.\"  All rights reserved.
.\"
.TH FIREWALK 8 "04.20.2002" "firewalk"
.SH NAME
firewalk \- Active Reconnaissance Network Security Tool with Extreme Prejudice
.SH SYNOPSIS
.B firewalk
[\-dhinprSsTtvx] target_gateway metric
.if n .ti +5n
.SH DESCRIPTION
Firewalk is an active reconnaissance network security tool that attempts to
determine what layer 4 protocols a  given IP forwarding device will pass.
Firewalk  works  by sending out TCP or UDP packets with a TTL one greater
than the targeted gateway.  If the gateway allows the traffic, it will
forward the packets to the next hop where they will expire and elicit 
an ICMP_TIME_EXCEEDED  message.  If the gateway hostdoes not allow the
traffic, it will likely drop the packets on  the floor and we will see no
response.

To get  the  correct  IP  TTL that will result in expired packets one beyond
the gateway we need  to  ramp  up  hop-counts.   We  do  this  in the same
manner that traceroute works.  Once we have the gateway hopcount (at  that
point the scan is said to be `bound`) we can begin our scan.

It is significant to note the fact that the ultimate destination host does not
have to be reached.  It just  needs to be somewhere downstream, on the other
side of the gateway, from the scanning host.  Please see http://www.wiley.com/cda/product/0,,0471205443,00.html
for more information on Firewalking and networking security tools in general.

.SH COMMAND-LINE OPTIONS
If an option takes an argument, it procedes the option letter, with the
default in parenthesis.
.TP 15
.B \-d 1-65535 (34434)
Specify the initial destination port to use during the network discovery
(aka TTL ramping) phase.
.TP
.B \-h
Program help.
.TP
.B \-i interface_name
Specify interface to use.  Only neccessary on multi-homed machines.
.TP
.B \-n
Do not resolve IP addresses into hostnames.  This saves a DNS lookup
and speeds the scans (mainly during network discovery).
.TP
.B \-P 1-2000 (0)
Set a network writing pause which may be neccessary to keep the program from
flooding the network.
.TP
.B \-p TCP, UDP (UDP)
Type of scan to perform.
.TP
.B \-r
Strict RFC 793 compliance.  This only comes into play when doing a TCP scan
when your packets have an expire vector of one and your metric host is one
hop from your gateway.  Since the packets will reach their destination, they
will not expire, so we look for terminal responses.  For a TCP port in the
listen state, we will get back a SYN|ACK with the ACK as our SEQ + 1.  However,
for a closed port, the response is stack dependent.  If the host is RFC
compliant we will receive an RST|ACK with the ACK as our SEQ + 1.  However, 
if the host is not compliant (ie: microsoft) then the best we can do is
inverse tuple matching (which is the default).
.TP
.B \-S 1-65535,... (1-130,139,1025)
Specify the ports for the scan.  Ports may be specified in ranges, delimited by
dashes, and multiple ranges may be specified, delimited by commas.  Ommiting
the terminating port number is shorthand for 65535.
.TP
.B \-s 1-65535 (53)
Specify the source port for the scan (both phases).
.TP
.B \-T 1-2000 (2)
Network packet reading timeout.  This is the time firewalk will spend
waiting for a response before timing out.
.TP
.B \-t 1-25 (1)
Set the initial IP time to live (TTL) value.  If a target gateway is known
to be (at least) n hops from the source host, the TTL can be preloaded to
facilitate a faster scan.
.TP
.B \-v
Dump program version and exit.
.TP
.B \-x expire vector (1)
The expire vector is the number of hops that the scanning probes will
expire, past the gateway host.  The binding hopcount is the hopcount of the
gateway + the expire vector.
.TP

.SH COMMAND-LINE EXAMPLES
.LP

.SH CAVEATS
.LP

.SH SEE ALSO
.LP
traceroute(8), tracerx(8), pcap(3), libnet(3), dnet(3)
.SH AUTHOR
.LP
Mike D. Schiffman <mike@infonexus.com>
.SH BUGS
.LP
Please send bug reports to mike@infonexus.com


Firewalk/src/firewalk.c
/*
 *  $Id: firewalk.c,v 1.2 2002/05/14 23:28:37 route Exp $
 *
 *  Firewalk 5.0
 *  firewalk.c - Scanning driver
 *
 *  Copyright (c) 1998 - 2002 Mike D. Schiffman  <mike@infonexus.com>
 *  Copyright (c) 1998, 1999 David E. Goldsmith <dave@infonexus.com>
 *  http://www.packetfactory.net/firewalk
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if (HAVE_CONFIG_H)
#include "../include/config.h"
#endif
#include "../include/firewalk.h"

int
firewalk(struct firepack **fp)
{
    int done, i, j;
    u_short bport, cport, eport;

    /* inform the user what's what */
    printf("%s-based scan.\n",
            (*fp)->protocol == IPPROTO_TCP ? "TCP" : "UDP");
    printf("Ramping phase source port: %d, destination port: %d\n",
            (*fp)->sport, (*fp)->dport);
    if ((*fp)->flags & FW_STRICT_RFC && (*fp)->protocol == IPPROTO_TCP)
    {
        printf("Using strict RFC adherence.\n");
    }
    printf("Hotfoot through %s using %s as a metric.\n",
            libnet_addr2name4(((*fp)->gateway),
            ((*fp)->flags) & FW_RESOLVE),
            libnet_addr2name4(((*fp)->metric),
            ((*fp)->flags) & FW_RESOLVE));

    /*
     *  PHASE ONE: Firewalk hopcount ramping
     *  A standard Traceroute-style IP expiry scan is initiated towards
     *  the metric, with the intent being to find how many hops away the
     *  target gateway is from the scanning host.  We'll increment the
     *  hopcounter and update packet template each pass through the loop.
     */
    printf("Ramping Phase:\n");
    for (done = 0, i = 0; !done && i < FW_IP_HOP_MAX; i++)
    {
        /* send a series of probes (currently only one) */
        for (j = 0; j < 1; j++)
        {
            fprintf(stderr, "%2d (TTL %2d): ", i + 1, (*fp)->ttl);
            if (fw_packet_inject(fp) == -1)
            {
                /*
                 *  Perhaps this write error was transient.  We'll hope
                 *  for the best.  Inform the user and continue.
                 */
                fprintf(stderr, "fw_packet_inject(): %s\n",
                        (*fp)->errbuf);
                continue;
            }
            switch (fw_packet_capture(fp))
            {
                case FW_PACKET_IS_UNREACH_EN_ROUTE:
                case FW_PACKET_IS_TTL_EX_EN_ROUTE:
                    if ((*fp)->flags & FW_BOUND)
                    {
                        printf("Binding host reached.\n");
                        done = 1;
                    }
                    break;
                case FW_PACKET_IS_TERMINAL_TTL_EX:
                case FW_PACKET_IS_TERMINAL_UNREACH:
                case FW_PACKET_IS_TERMINAL_SYNACK:
                case FW_PACKET_IS_TERMINAL_RST:
                    /* any terminal response will end phase one */
                    done = 1;
                    break;
                case -1:
                case FW_SERIOUS_ERROR:
                    /* err msg set in fw_packet_capture() */
                    return (FW_SERIOUS_ERROR);
                case FW_USER_INTERRUPT:
                    /* user hit ctrl-c */
                    return (FW_USER_INTERRUPT);
            }
        }
        if (!done)
        {
            if (fw_packet_update_probe(fp, 0) == -1)
            {
                /* error msg set in fw_packet_update_probe */
                return (-1);
            }
        }
    }
    if (done && !((*fp)->flags & FW_BOUND))
    {
        /*
         *  If we're "done" but not "bound" then we hit the metric
         *  before we hit the target gateway.  This means the target
         *  gateway is not en route to the metric.  Game's over kids.
         */
        sprintf((*fp)->errbuf, 
                "metric responded before target; must not be en route");
        return (FW_ABORT_SCAN);
    }
    if (!done)
    {
        /* if we fall through down here, we've exceeded our hopcount */
        sprintf((*fp)->errbuf, "hopcount exceeded");
        return (FW_ABORT_SCAN);
    }

    /*
     *  PHASE TWO: Firewalk scanning
     *  A series of probes are sent from to the metric with the bound IP
     *  TTL. If a given probe is accepted through the target gateway's
     *  ACL, we will receive an ICMP TTL expired in transit from the
     *  binding host If we receive no response after the timeout expires,
     *  it is assumed the probe violated the ACL on the target and was
     *  dropped.
     */
    (*fp)->ttl += (*fp)->xv;
    printf("Scan bound at %d hops.\n", (*fp)->ttl);
    printf("Scanning Phase: \n");
    for (done = 0, i = 0; !done; i++)
    {
        if (!libnet_plist_chain_next_pair((*fp)->plist, &bport, &eport))
        {
            /* we've exhausted our portlist and we're done */
            done = 1;
            continue;
        }
        while (!(bport > eport) && bport != 0)
        {
            cport = bport++;
            if (fw_packet_update_probe(fp, cport) == -1)
            {
                /* error msg set in fw_packet_update_probe */
                return (-1);
            }

            /* send a series of probes (currently only one) */
            for (j = 0; j < 1; j++)
            {
                fprintf(stderr, "port %3d: ", cport);
                (*fp)->stats.ports_total++;
                if (fw_packet_inject(fp) == -1)
                {
                    /*
                     *  Perhaps this write error was transient.  We'll
                     *  hope for the best.  Inform the user and continue.
                     */
                    fprintf(stderr, "fw_packet_inject(): %s\n",
                            (*fp)->errbuf);
                    continue;
                }
                /* we don't care what the return value is this time */
                switch(fw_packet_capture(fp))
                {
                    case FW_USER_INTERRUPT:
                        return (FW_USER_INTERRUPT);
                    case -1:
                    case FW_SERIOUS_ERROR:
                        /* err msg set in fw_packet_capture() */
                        return (FW_SERIOUS_ERROR);
                    default:
                        /* empty */
                }
            }
        }
    }
    return (1);
}

/* EOF */
